consensus/src/reconstruction/mod.rs

```diff
@@ -0,0 +1,211 @@
+        &mut self,
+        reconstruction_result: ReconstructionResult<H>,
+    ) -> Vec<Output<H>> {
+        use Output::*;
+        let mut result = Vec::new();
+        let (units, requests) = reconstruction_result.into();
+        result.append(&mut requests.into_iter().map(Request).collect());
+        for unit in units {
+            result.append(&mut self.dag.add_unit(unit).into_iter().map(Unit).collect());
+        }
+        result
+    }
+
+    fn handle_notification(&mut self, notification: NotificationIn<H>) -> Vec<Output<H>> {
+        use NotificationIn::*;
+        let mut result = Vec::new();
+        match notification {
+            NewUnits(units) => {
+                for unit in units {
+                    let reconstruction_result = self.reconstruction.add_unit(unit);
+                    result.append(&mut self.handle_reconstruction_result(reconstruction_result));
+                }
+            }
+            UnitParents(unit, parents) => {
+                let reconstruction_result = self.reconstruction.add_parents(unit, parents);
+                result.append(&mut self.handle_reconstruction_result(reconstruction_result))
+            }
+        }
+        result
+    }
+
+    fn handle_output(&mut self, output: Output<H>) -> bool {
+        use Output::*;
+        match output {
+            Request(request) => {
+                if self
+                    .notifications_for_runway
+                    .unbounded_send(request.into())
+                    .is_err()
+                {
+                    warn!(target: LOG_TARGET, "Notification channel should be open.");
+                    return false;
+                }
+            }
+            Unit(unit) => {
+                // TODO add other destinations
+                if self.units_for_creator.unbounded_send(unit.unit()).is_err() {
+                    warn!(target: LOG_TARGET, "Creator channel should be open.");
+                    return false;
+                }
+                if self
+                    .units_for_extender
+                    .unbounded_send(unit.extender_unit())
+                    .is_err()
+                {
+                    warn!(target: LOG_TARGET, "Extender channel should be open.");
+                    return false;
+                }
+                if self
+                    .notifications_for_runway
+                    .unbounded_send(NotificationOut::AddedToDag(
+                        unit.hash(),
+                        unit.parents().clone().into_values().collect(),
+                    ))
+                    .is_err()
+                {
+                    warn!(target: LOG_TARGET, "Notification channel should be open.");
+                    return false;
+                }
+            }
+        }
+        true
+    }
+
+    /// Run the reconstruction service until terminated.
+    pub async fn run(mut self, mut terminator: Terminator) {
+        let mut exiting = false;
```

Comment:
Hmm, I would prefer to avoid using labels and at least one of the assignments is in a nested loop (not sure about the others, doesn't `futures::select` expand into a loop?). I could factor out the select into a method returning whether to continue, would that be better? Note that this flag is from a pattern that we iirc use also in other places, so there is also a bit of a cost to less consistency (not much in this case tho).

Reply:
```suggestion
            /// Run the reconstruction service until terminated.
    pub async fn run(mut self, mut terminator: Terminator) {
        loop {
            futures::select! {
                n = self.notifications_from_runway.next() => match n {
                    Some(notification) => for output in self.handle_notification(notification) {
                        if !self.handle_output(output) {
                            break;
                        }
                    },
                    None => {
                        warn!(target: LOG_TARGET, "Notifications for reconstruction unexpectedly ended.");
                        break;
                    }
                },
                _ = terminator.get_exit().fuse() => {
                    debug!(target: LOG_TARGET, "Received exit signal.");
                    break;
                }
            }
        }
        debug!(target: LOG_TARGET, "Reconstruction decided to exit.");
        terminator.terminate_sync().await;
    }

```

Reply:
The first break only breaks out of the internal `for`, not the external `loop`.

Reply:
I missed that. You could also just `return` there (and when you see that `None`), which should be actually more consistent with `terminator''s semantics, i.e. if you call `terminate_sync()` before a termination call from parent, then the whole termination logic is skipped anyway.

Reply:
Done.